staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,750 @@
+// use.  We do this before holding over old priority levels so that
+// requests stop going to those levels and FlowSchemaStatus values
+// reflect this.  This function also adds any missing mandatory
+// FlowSchema objects.  The given objects must all have disstinct
+// names.
+func (meal *cfgMeal) digestFlowSchemasLocked(newFSs []*fctypesv1a1.FlowSchema) {
+	fsSeq := make(apihelpers.FlowSchemaSequence, 0, len(newFSs))
+	fsMap := make(map[string]*fctypesv1a1.FlowSchema, len(newFSs))
+	for i, fs := range newFSs {
+		otherFS := fsMap[fs.Name]
+		if otherFS != nil {
+			// This client is forbidden to do this.
+			panic(fmt.Sprintf("Given two FlowSchema objects with the same name: %#+v and %#+v", fcfmt.Fmt(otherFS), fcfmt.Fmt(fs)))
+		}
+		fsMap[fs.Name] = fs
+		if fs.Name == tracerName {
+			meal.cfgCtl.setTracedValueLocked(fs.Spec.MatchingPrecedence)
+		}
+		_, goodPriorityRef := meal.newPLStates[fs.Spec.PriorityLevelConfiguration.Name]
+
+		// Ensure the object's status reflects whether its priority
+		// level reference is broken.
+		//
+		// TODO: consider not even trying if server is not handling
+		// requests yet.
+		meal.presyncFlowSchemaStatus(fs, !goodPriorityRef)
+
+		if !goodPriorityRef {
+			klog.V(6).Infof("Ignoring FlowSchema %s because of bad priority level reference %q", fs.Name, fs.Spec.PriorityLevelConfiguration.Name)
+			continue
+		}
+		fsSeq = append(fsSeq, newFSs[i])
+		meal.haveExemptFS = meal.haveExemptFS || fs.Name == fctypesv1a1.FlowSchemaNameExempt
+		meal.haveCatchAllFS = meal.haveCatchAllFS || fs.Name == fctypesv1a1.FlowSchemaNameCatchAll
+	}
+	// sort into the order to be used for matching
+	sort.Sort(fsSeq)
+
+	// Supply missing mandatory FlowSchemas, in correct position
+	if !meal.haveExemptFS {
+		fsSeq = append(apihelpers.FlowSchemaSequence{fcboot.MandatoryFlowSchemaExempt}, fsSeq...)
+	}
+	if !meal.haveCatchAllFS {
+		fsSeq = append(fsSeq, fcboot.MandatoryFlowSchemaCatchAll)
+	}
+
+	meal.cfgCtl.flowSchemas = fsSeq
+	if klog.V(5) {
+		for _, fs := range fsSeq {
+			klog.Infof("Using FlowSchema %#+v", fcfmt.Fmt(fs))
+		}
+	}
+}
+
+// Consider all the priority levels in the previous configuration.
+// Keep the ones that are in the new config, supply mandatory
+// behavior, or are still busy; for the rest: drop it if it has no
+// queues, otherwise start the quiescing process if that has not
+// already been started.
+func (meal *cfgMeal) processOldPLsLocked() {
+	for plName, plState := range meal.cfgCtl.priorityLevelStates {
+		if meal.newPLStates[plName] != nil {
+			// Still desired and already updated
+			continue
+		}
+		if plName == fctypesv1a1.PriorityLevelConfigurationNameExempt && !meal.haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll && !meal.haveCatchAllPL {
+			klog.V(3).Infof("Retaining mandatory priority level %q despite lack of API object", plName)
+		} else {
+			if plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle() {
```

Comment:
Replacing `plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle()` with `plState.numPending == 0 && plState.queues.IsIdle()` would make this code remove a priority level any time it is empty --- which would be wrong.  We only want to remove a priority level when it is empty _and_ it does not have a corresponding API object.

Reply:
Huh? We've already exited the loop early if the newPLStates[plName] is non-nil (i.e. if the corresponding API object exists). But we *haven't* already set plState.quiescing if the API object does not exist, so this will mean we always have to go through at least two rounds of config changes to drop a PL.

Reply:
Oh, right, checking `plState.quiescing` at this point _is_ redundant.  Good catch.

Reply:
removed